File: leet-code/sortedArrayToBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
function TreeNode (val) {
  this.val = val
  this.left = this.right = null
}
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function (nums) {
  return generateTree(nums, 0, nums.length - 1)
}

function generateTree (arr, left, right) {
  if (left > right) {
    return null
  }
  let mid = parseInt((right + left) / 2)
  console.log(left, right, mid)
  let node = new TreeNode(arr[mid])
  console.log(mid)
  node.left = generateTree(arr, left, mid - 1)
  node.right = generateTree(arr, mid + 1, right)
  return node
}

console.log(sortedArrayToBST([-10, -3, 0, 5, 9]))

File: leet-code/sortedArrayToBST.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
	return generateTree(nums, 0, len(nums)-1)
}

func generateTree(nums []int, left int, right int) *TreeNode {
	if left > right {
		return nil
	}
	var node *TreeNode = new(TreeNode)
	mid := (left + right) / 2
	node.Val = nums[mid]
	node.Left = generateTree(nums, left, mid-1)
	node.Right = generateTree(nums, mid+1, right)
	return node
}
